File: 107.binary-tree-level-order-traversal-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	public List<List<Integer>> levelOrderBottom(TreeNode root) {
		List<List<Integer>> result = new ArrayList<List<Integer>>();
		visit(root,result,0);
		Collections.reverse(result);
		return result;       
	}

	public void visit(TreeNode node, List<List<Integer>> list, int level){
		if (node  == null) return;
		if(list.size() < level + 1)  list.add(new ArrayList<Integer>()); //这个可以通过最顶层的root 想出来，list.size 现在为空，但是已经有一层了，所以要new
		list.get(level).add(node.val);
		visit(node.left,list,level+1);
		visit(node.right,list,level+1);
	}
}

File: 107.binary-tree-level-order-traversal-ii.js
/*
 * @lc app=leetcode id=107 lang=javascript
 *
 * [107] Binary Tree Level Order Traversal II
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
// var levelOrderBottom = function (root) {
//   let result = [];
//   if (!root) return result;
//   function helper(root, level) {
//     if (!root) return;
//     helper(root.left, level + 1);
//     helper(root.right, level + 1);
//     if (!result[level]) {
//       result[level] = new Array();
//     }
//     result[level].push(root.val);
//   }

//   helper(root, 0);
//   return result.reverse();
// };


// BFS 来做可能会比较好
var levelOrderBottom = function (root) {
  let result = [];
  if (!root) return result;
  let queue = [root];
  while (queue.length) {
    let size = queue.length;
    let temp = new Array();
    for (let i = 0; i < size; i++) {
      let node = queue.shift();
      temp.push(node.val);
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right)
    }
    result.unshift(temp);

  }
  return result;
};

